### ISSUE TITLE
API support

### ISSUE THREAD

**Issue by dominicbr** [ShaneMcC/PXE-Manager#21]
Hi Shane,

Please can you look at adding API support to allow automated creation of server deployment images.

Ideally authentication would be required/supported.

Thanks!
Dom

**Comment by ShaneMcC** [comment]
Hi Dom - thanks for the issue. It's been sorta on my todo list for a while so :)

I'll have a look at what is needed here when I get some free time. A lot of the data for the frontend is already manipulated via ajax and json anyway, so it shouldn't be too difficult, a few extra json end points for authentication and GETing data.

**Comment by ShaneMcC** [comment]
*Very* rough support is now available for entirely programmatic interaction as long as you are using one of the default login-based auth providers.

If NullAuthProvider is in use, then this obviously won't work, the others all should.

Need to get it documented.

At the moment, because there's no proper separate api endpoint it's a bit rough to use as it's still really just a website masqurading as an API by providing the data as JSON not HTML.

I've versioned this as api 0.1 for now, so when I redo this in future to be nicer and not a hack I can try to keep backwards-compatability.

TL;DR:

- GET /api/0.1/user.json to learn your CSRFTOKEN which is needed for ALL 'POST' requests and what your current autentication state is and what permissions you have.
- if not authenticated or you don't have the required permissions.
  - (Optionally) GET /api/0.1/login.json to learn what fields are required to be POSTed (For most people this is entirely optional as you'll know what auth provider you have enabled and what fields it requires)
  - POST to /login.json something like: ```{"pass": "admin123", "csrftoken": "E49B6FCB-7D73-4AB4-A97F-1C79B463BCA5"}```
  - The returned data will include an `"error"` or ` "success"`  field as a result. If success, you will get a session cookie set, and also get a `"sessionid"` field in the output. Either this cookie, or an `X-SESSION-ID` header is required on all future requests so that authentication can be checked.
- GET/POST to the various /servers and /images endpoints as required.

The following endpoints exist:
**GET** `/api/0.1/user.json` - Get current user details
**GET** `/api/0.1/login.json` - Get information required for login
**POST** `/api/0.1/login.json` - Login
**GET** `/api/0.1/logout.json` - Logout

**GET** `/api/0.1/servers.json` - Get information about all servers
**GET** `/api/0.1/servers/mac/<macaddress.json` - Get information about server with the given mac address
**GET** `/api/0.1/servers/<serverid>.json` - Get information about server with the given ID
**GET** `/api/0.1/servers/<serverid>/preview.json` - Get preview data for server with the given ID
**POST** `/api/0.1/servers/<serverid>/delete.json` - Delete server with the given ID
**POST** `/api/0.1/servers/<serverid>/duplicate.json` - Duplicate server with the given ID
**POST** `/api/0.1/servers/create.json` - Create a new server
**POST** `/api/0.1/servers/<serverid>/edit.json` - Edit server with the given ID

**GET** `/api/0.1/images.json` - Get information about all images
**GET** `/api/0.1/images/available.json` - Get information about all available images
**GET** `/api/0.1/images/<imageid>.json` - Get information about image with the given ID
**POST** `/api/0.1/images/<imageid>/duplicate.json` - Duplicate image with the given ID
**GET** `/api/0.1/images/<imageid>/requiredvariables.json` - Get all the required variables for an image with the given ID (Including those from parents)
**POST** `/api/0.1/images/create.json` - Create a new image
**POST** `/api/0.1/images/<imageid>/edit.json` - Edit image with the given ID
**POST** `/api/0.1/images/<imageid>`/delete.json` - Delete with the given ID

For all POST endpoints, data can be either JSON or form-encoded (If it parses as a valid json array, it will be treated as json else as form-encoded), and must include a csrftoken.

The fields for each end point will be documented separately in future (but for now can be worked out by looking at devtools as the ui POSTs to the same end points)

(Technically, right now, The "/api/0.1/" bit is an optional hack, but I'm not commiting to keeping backwards-compatability for unversioned endpoints)

**Comment by ShaneMcC** [comment]
Future work on this should involve:
- Documenting each endpoint
- Allowing inline-authentication for each endpoint rather than needing to obtain a session token first

**Comment by ShaneMcC** [comment]
Closing this issue.

At this time I'm unlikely to spend too much time working more on the API side of things. I won't break it though so it will continue to work like now.

I've added a wiki page pointing at this issue for documentation for now, maybe I'll document it properly in future.
